src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,276 @@
+#pragma once
+
+#include <fmt/format.h>
+#include <Storages/MergeTree/RequestResponse.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+/// A boundary of a segment (left or right)
+struct PartToRead
+{
+    PartBlockRange range;
+    String name;
+
+    bool operator==(const PartToRead & rhs) const
+    {
+        return range == rhs.range && name == rhs.name;
+    }
+
+    struct FindByBegin : std::binary_function<bool, Int64, PartToRead>
+    {
+        bool operator()(Int64 value, const PartToRead & rhs) const
+        {
+            return value < rhs.range.begin;
+        }
+    };
+
+    struct FindByEndGreater : std::binary_function<bool, Int64, PartToRead>
+    {
+        bool operator()(Int64 value, const PartToRead & lhs) const
+        {
+            return value > lhs.range.end;
+        }
+    };
+
+    struct CompareByBegin : std::binary_function<bool, PartToRead, PartToRead>
+    {
+        bool operator()(const PartToRead & lhs, const PartToRead & rhs) const
+        {
+            return lhs.range.begin < rhs.range.begin;
+        }
+    };
+
+
+    struct CompareByEnd : std::binary_function<bool, PartToRead, PartToRead>
+    {
+        bool operator()(const PartToRead & lhs, const PartToRead & rhs) const
+        {
+            return lhs.range.end < rhs.range.end;
+        }
+    };
+};
+
+
+class PartRangesIntersectionsIndex
+{
+public:
+    void addPart(PartToRead part)
+    {
+        by_begin.insert(part);
+        by_end.insert(part);
+    }
+
+    size_t numberOfIntersectionsWith(PartBlockRange range)
+    {
+        /// Find the first one that starts with a larger coordinate
+        auto right_iter = std::upper_bound(by_begin.begin(), by_begin.end(), range.end, PartToRead::FindByBegin());
```

Comment:
This is linear, because set iterator is not random access
see https://en.cppreference.com/w/cpp/algorithm/upper_bound